// @esri/calcite-components/dist/cjs/locale-69dc9583.js
/*!
 * All material copyright ESRI, All Rights Reserved, unless otherwise specified.
 * See https://github.com/Esri/calcite-components/blob/master/LICENSE.md for details.
 */
'use strict';

const number = require('./number-e6448e15.js');

const locales = [
  "ar",
  "bs",
  "ca",
  "cs",
  "da",
  "de",
  "de-CH",
  "el",
  "en",
  "en-AU",
  "en-CA",
  "en-GB",
  "es",
  "es-MX",
  "et",
  "fi",
  "fr",
  "fr-CH",
  "he",
  "hi",
  "hr",
  "hu",
  "id",
  "it",
  "it-CH",
  "ja",
  "ko",
  "lt",
  "lv",
  "mk",
  "nb",
  "nl",
  "pl",
  "pt",
  "pt-PT",
  "ro",
  "ru",
  "sk",
  "sl",
  "sr",
  "sv",
  "th",
  "tr",
  "uk",
  "vi",
  "zh-CN",
  "zh-HK",
  "zh-TW"
];
function createLocaleNumberFormatter(locale) {
  return new Intl.NumberFormat(locale, {
    minimumFractionDigits: 0,
    maximumFractionDigits: 20
  });
}
function delocalizeNumberString(numberString, locale) {
  if (numberString) {
    const groupSeparator = getGroupSeparator(locale);
    const decimalSeparator = getDecimalSeparator(locale);
    const splitNumberString = numberString.split("");
    const decimalIndex = splitNumberString.lastIndexOf(decimalSeparator);
    const delocalizedNumberString = splitNumberString
      .map((value, index) => {
      if (value === groupSeparator || (value === decimalSeparator && index !== decimalIndex)) {
        return "";
      }
      return value;
    })
      .reduce((string, part) => string + part)
      .replace(decimalSeparator, ".");
    return isNaN(Number(delocalizedNumberString)) ? numberString : delocalizedNumberString;
  }
  return numberString;
}
function getGroupSeparator(locale) {
  const formatter = createLocaleNumberFormatter(locale);
  const parts = formatter.formatToParts(1234567.8);
  const value = parts.find((part) => part.type === "group").value;
  return value.trim().length === 0 ? " " : value;
}
function getDecimalSeparator(locale) {
  const formatter = createLocaleNumberFormatter(locale);
  const parts = formatter.formatToParts(1234567.8);
  const value = parts.find((part) => part.type === "decimal").value;
  return value.trim().length === 0 ? " " : value;
}
function localizeNumberString(numberString, locale, displayGroupSeparator = false) {
  if (numberString) {
    const number$1 = Number(number.sanitizeDecimalString(numberString));
    if (!isNaN(number$1)) {
      const formatter = createLocaleNumberFormatter(locale);
      const parts = formatter.formatToParts(number$1);
      const localizedNumberString = parts
        .map(({ type, value }) => {
        switch (type) {
          case "group":
            return displayGroupSeparator ? getGroupSeparator(locale) : "";
          case "decimal":
            return getDecimalSeparator(locale);
          default:
            return value;
        }
      })
        .reduce((string, part) => string + part);
      return localizedNumberString;
    }
  }
  return numberString;
}

exports.delocalizeNumberString = delocalizeNumberString;
exports.getDecimalSeparator = getDecimalSeparator;
exports.locales = locales;
exports.localizeNumberString = localizeNumberString;
